### ISSUE TITLE
Custom virtual sites

### ISSUE THREAD

**Issue by selimsami** [openmm/openmm#4296]
Hi,

Is there a way to add custom virtual sites, just like you add custom forces? If not, could you point out which parts of the code I would need to change to add my custom virtual site?

To give you an example, I would like the virtual site placement to depend on the r_ij bond vector times the deviation from the equilibrium bond lengths (r_ij-r0) on top of the common virtual site placement rules like weighted center of geometry.

(sounds a little crazy, I know)

Thanks.
Selim

**Comment by peastman** [comment]
The most general virtual site type is [LocalCoordinatesSite](http://docs.openmm.org/latest/api-c++/generated/LocalCoordinatesSite.html).  It doesn't let you specify arbitrary functions, but it does have a fairly flexible form.  A fully custom virtual site would be challenging to implement, since the calculation has to be invertible (to propagate forces on the virtual site back to the original particles).

**Comment by selimsami** [comment]
Right, thought so. LocalCoordinatesSite unfortunately won't work for what I have in mind.
Could you perhaps point me at the right direction for implementing my own virtual site placement & force propagation?
Thanks

**Comment by peastman** [comment]
That could be challenging depending on what you have in mind, but you're certainly welcome to try!  The first step is to add the API for it in VirtualSite.h and VirtualSite.cpp.  Next you need to write the calculations for it.  Start with ReferenceVirtualSites.cpp.  That's the implementation used by the Reference and CPU platforms.  Add another block in `computePositions()` that computes the position for your new virtual site class, and another block in `distributeForces()` that propagates the forces from it.

You'll then need to implement it for the CUDA and OpenCL platforms, which is a little more complicated, but start with the reference version.  If you can get that working, the other platforms will be easy.

**Comment by selimsami** [comment]
That's very clear, thank you. I will give it a try and update here.

**Comment by peastman** [comment]
The CPU platform always uses mixed precision.  If that's good enough for your purposes, you can start testing it.

To implement the CUDA and OpenCL versions (they're both implemented with the same code), there are two parts.  First, there's C++ code that records the parameters for all the virtual sites and uploads them to the GPU.  It starts at [this line](https://github.com/openmm/openmm/blob/796ffaaa97decdf1593ed75afa5fbb21b6b32e9c/platforms/common/src/IntegrationUtilities.cpp#L422).  You should be able to add the code for your virtual site class by copying what the other classes do.

Next you need to write the GPU code.  That means updating `computeVirtualSites()` and `distributeVirtualSiteForces()`, both found in [integrationUtilities.cc](https://github.com/openmm/openmm/blob/master/platforms/common/src/kernels/integrationUtilities.cc).  This again should be straightforward to do by just copying what the other classes do, even if you haven't programmed GPUs before.

**Comment by selimsami** [comment]
Thanks a lot @peastman! That seems quite straightforward. I also noticed some virtual site code in [ComputeContext.cpp](https://github.com/openmm/openmm/blob/796ffaaa97decdf1593ed75afa5fbb21b6b32e9c/platforms/common/src/ComputeContext.cpp), is this something I should also change?

A bit off-topic to this issue, but from my initial tests, the performance I get is quite a bit worse than I expected for my Drude-based force field. For a box of 216 water molecules, I get about 22 seconds per step in the Reference platform and 1.6 seconds per step in CPU platform with 6 cores on a macbook pro (same performance with or without my new virtual site). Not a fair comparison because I was using tabulated potentials there, but in GROMACS I was getting about 0.04 seconds per step single core and double precision.
I guess my question is, would you expect the use of tabulated potentials to make such a big difference? If not, would you mind having a quick look at my code here or in a different issue to see if I'm doing something inefficiently?

**Comment by peastman** [comment]
> I also noticed some virtual site code in ComputeContext.cpp, is this something I should also change?

Yes, sorry I missed that.

> For a box of 216 water molecules, I get about 22 seconds per step in the Reference platform

There's definitely something wrong in there.  It's hard to guess what.  The first thing to do is test the standard OpenMM code without your new virtual site class.  That will tell us whether the problem is in your code or somewhere else.  And of course running it in a profiler to see where the time is spent would be useful.

**Comment by selimsami** [comment]
@peastman, I have just tried with a conda installed openmm, and there are no changes in the performance.

does openmm provide time accounting after a simulation? I did not manage to find something like this. If not, do you have a suggestion for the profiling? I have mostly experience with python profilers.

I attach my force field in case you notice something obviously inefficient. I have a somewhat complicated functional form - so I except some of the extra cost to be due to this.

[openmm_test.zip](https://github.com/openmm/openmm/files/13653629/openmm_test.zip)

**Comment by peastman** [comment]
Is there a reason you're using DrudeSCFIntegrator?  DrudeLangevinIntegrator is *much* faster.  The SCF integrator needs to run an energy minimization at every step to solve for the Drude particle positions.

**Comment by peastman** [comment]
To give a sense of just how much overhead you're adding, I compared integrating to just computing forces:

```python
time_begin = time.time()
simulation.step(5)
time_end = time.time()
print('simulation time', time_end-time_begin)

time_begin = time.time()
for i in range(5):
    simulation.context.getState(getForces=True)
time_end = time.time()
print('force evaluation time', time_end-time_begin)
```

Here is the output:

```
simulation time 7.338864088058472
force evaluation time 0.11242794990539551
```

**Comment by selimsami** [comment]
Mostly because I wanted to compare it to my GROMACS implementation, I was eventually planning to switch to Langevin. But wow, I did not expect that much difference! 

I also did not expect the SCF implementation to be this slow though. Are core-core interactions automatically skipped during the SCF procedure?

Also, is 7.33 seconds for the code I have shared with you, or have you changed something? Because that takes 113 seconds for me with the reference platform.

**Comment by peastman** [comment]
That was with the CPU platform.  The SCF method is inherently much slower than Langevin, but the implementation in OpenMM is particularly slow.  It just uses the generic energy minimizer.  I should try to speed it up one of these days.

**Comment by selimsami** [comment]
Ah I see that explains it. The difference is between SCF and Lagrangian [is not supposed to be ](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4481176/)that big indeed. I find the  [GROMACS](https://gitlab.com/gromacs/gromacs/-/blob/main/src/gromacs/mdrun/shellfc.cpp#L645) implementation rather simple and quite fast. A good shell optimizer is useful I think regardless of the advantage of the extended Lagrangian in most cases. For example, I do my parameter optimization on dimer/pentamer clusters, for which I only need to optimize the shells. Since I need tens of thousands of these calculations (large dataset + parameter optimization), it is extremely slow to do that in openmm.


Sorry for taking so much of your time, I have two more questions:

1) With Langevin, you need frictionCoeff, drudeTemperature, drudeFrictionCoef. Is there default or commonly used values for these? I did not manage to find any information about that.  (site note: I guess smaller time-step requirements (or multi-stepping) reduces some of the advantages of Lagrangian compared to your above example)

2) Even with your example, a single force call takes 0.02 seconds with 4-6 cores. GROMACS single core double precision single force call takes 0.003 seconds. That's still quite a difference. Would you guess it's mostly due to tabulated potentials?  I guess I could try using tabulated potentials in openmm too, would that make sense?

**Comment by peastman** [comment]
Some reasonable defaults would be 1/ps for the friction coefficient and maybe 10 K for the Drude temperature.  I'm not sure what's typically used for the Drude friction coefficient.  I would guess you want it a bit higher than the regular friction, maybe 10/ps, but that's just a guess.

Keep in mind you have a ton of custom forces, including six different CustomNonbondedForces.  On the CUDA or OpenCL platform they all get compiled into a single kernel, so it's very fast.  On CPU they all get compiled and executed separately, so there's more overhead for each one.

**Comment by peastman** [comment]
I just wrote a special purpose minimizer for DrudeSCFIntegrator, and it brought the simulation time down from 7.33 seconds to only 1.17 seconds.  Still a lot slower than Langevin, but a huge improvement.  I'll do the same for the GPU platforms and get a PR in shortly.

**Comment by selimsami** [comment]
Oh that's great, thanks a lot for that. Looking forward to giving it a try.

Another advantage of having a reasonably fast DrudeSCF is actually so that you can test to those DrudeLangevin parameters to see if you're getting the same results and choose correct ones.

I just tested it on OpenCL (with standard virtual sites) and it was about 6 times faster compared to CPU (although not a great comparison: single vs mixed precision  + GPU instead of CPU). Combining that with the new DrudeSCF, that should already be quite good!


Also removing the following line made everything almost twice faster as well (which seems to be the case for PME only):
`system.setDefaultPeriodicBoxVectors([20, 0, 0], [0, 20, 0], [0, 0, 20])`
I guess some PME parameters are initialized based on this, even though I change the box size later to a much smaller value?

**Comment by peastman** [comment]
Right, the grid size is chosen at the start of the simulation.  It doesn't change even if the box size changes later.  It's assumed that variations in the box size will be fairly small, and changing the grid size would discontinuously change the energy.

**Comment by peastman** [comment]
I found another reason why it's so slow: the minimization tolerance is way too low.  You set it to 0.01 kJ/mol/nm, which is far smaller than it can achieve.  Even the default value 0.1 kJ/mol/nm is far too low.  We should consider increasing it.  The optimizer thrashes for a long time before finally giving up, usually with a RMS force on the Drude particles somewhere between 10 and 50 kJ/mol/nm.  If you set the tolerance to 50 it gets much faster with no change in the actual accuracy.

**Comment by selimsami** [comment]
That's interesting, but this might be connected to the DrudeSCF optimizer though.
In GROMACS I am able to get < 0.1 kJ/mol/nm rms force in 12 SCF steps.
And it is able to converge all the way down to 1e-10 rmsF in 30 steps.

That's actually another thing that I found could be interesting to add here:
- An option to specify the maximum number of SCF steps allowed - so that it stops the SCF and continues dynamics even if it does not reach convergence (perhaps with a warning). So that it does not get stuck and go on trying to optimize forever.
- (maybe I missed this but) the ability to see how many SCF steps it takes and the force/energy at each of these SCF steps would also be useful to see.

**Comment by peastman** [comment]
The new algorithm is in #4354.  If you want to give it a try, that would be great.

> And it is able to converge all the way down to 1e-10 rmsF in 30 steps.

Either they're measuring something different, or they're working entirely in double precision.  There's no way you'll get it anywhere close to that in single precision.

**Comment by selimsami** [comment]
Thanks I'll definitely give it a try.

That was double precision.
I just tried it in single precision and it can only go down to 0.04 kJ/mol/nm

**Comment by selimsami** [comment]
I'm not sure if it's a good idea to increase to convergence criteria so much btw. 
In DrudeSCF you're under the assumption that forces on drude particles are (almost) zero since you're doing the dynamics only other core particles.
Your energy conservation will be quite bad if you have large forces on the Drude particles.

**Comment by peastman** [comment]
How about 1 kJ/mol/nm?  Testing on your system, I find that if everything is done in double precision it can get the forces down to 0.1.  But in single precision mode, it can't.  It tries and eventually gives up with them somewhere around 0.2.

**Comment by selimsami** [comment]
As far as I know the forces should be a lot smaller than that. But I might be remembering something wrong.
I'm quite puzzled why openmm and gromacs force convergence would be so different.
I'll have a closer look tomorrow to see if they're both reporting similar things.

Also I wouldn't determine openmm defaults based on my untested new force field, maybe CHARMM drude FFs would be a better choice.

I started testing the PR, I also got a similar speedup for the runs.
One thing is, how does one make use of the new optimizer if not running dynamics?
For example if I want to do a geometry optimization, or just shell optimization?
simulation.minimizeEnergy() does not seem to use the new optimizer.
shouldn't there be something like a simulation.minimizeShells() that makes use of your new optimizer also?

**Comment by peastman** [comment]
You can estimate what level of precision should be achievable by looking at the magnitude of the harmonic restraining force.  For most of the Drude particles in your system, it's around 15,000 kJ/mol/nm.  It has to be precisely adjusted to cancel an equal force coming from other sources.  Each time you modify a particle position, you change both the restoring force and also all the other forces acting on it and on nearby particles.  Each of those forces is the result of a complicated series of operations, each of which introduces error at the level of machine epsilon.  In single precision that's about 5e-8.  If you want the result to be smaller than 0.01 kJ/mol/nm, you're asking for a relative accuracy of about 6e-7.  That means no force can have a relative error larger than about 10x machine epsilon.  That would be very difficult to achieve.

If you want to do geometry optimization, use the energy minimizer.  If you want to optimize just the Drude particles while keeping everything else fixed, I can see it might be useful to add a function for that.  You could do it by setting the step size to 0 and taking a step, but that's a little clumsy.

**Comment by selimsami** [comment]
Thanks for the explanation, that's very illuminating.

> If you want to do geometry optimization, use the energy minimizer. If you want to optimize just the Drude particles while keeping everything else fixed, I can see it might be useful to add a function for that. You could do it by setting the step size to 0 and taking a step, but that's a little clumsy.

Interesting. In GROMACS, Drude particles are fully optimized at each step of the geometry optimization. I wonder if there's an advantage of doing one over the other. Connected to that, what I find counterintuitive in openMM is that you have to give some mass to Drude particles in the DrudeSCF in order to minimize their position with the energy minimizer and dynamics. I think that should be a extended Lagrangian only requirement. 

An issue I am having actually might be connected to this (or I might be missing something):

When I initialize velocities:
`simulation.context.setVelocitiesToTemperature(300*kelvin)`

Both cores and shells are initialized to same velocities - still confused why DrudeSCF particles need velocities.
Then if I check the temperature with the reporter, it shows ~150 K.
If I compute the temperature myself from the Kinetic energy considering 3 particles per water, the temperature is ~300 K. So I guess it's counting the Drude particles to compute the temperature? do you know if this is  just a printing errror? are velocities correct?

Then it gets even more confusing when I add thermostat and start dynamics:
`system.addForce(AndersenThermostat(300*kelvin, 1/picosecond))`

Throughout at 500 ps run, the reported temperature is ~270 K. Never goes above 290 K (printing every few steps)

But if I compute the temperature myself again from the total kinetic energy considering 3 particles per water, then as you can expect the temperature is more like ~540K.

So I am quite confused if I am doing something wrong, whether it's just a printing issue, or the dynamics are happening at the wrong temperature.

**Comment by peastman** [comment]
It looks like that's a bug: DrudeLangevinIntegrator and DrudeNoseHooverIntegrator both implement `computeSystemTemperature()` to compute the temperature correctly, but DrudeSCFIntegrator doesn't.  We should fix that.

> In GROMACS, Drude particles are fully optimized at each step of the geometry optimization.

That's exactly what LocalEnergyMinimizer does.  It optimizes everything together.  There's no need to separate out the Drude particles in that case.

**Comment by peastman** [comment]
Actually, not a bug.  The assumption is that the Drude particles will be massless when using SCF.  Therefore they don't affect the temperature calculation.

The confusing thing here is that minimization doesn't work correctly when they're massless.  Drude particles aren't any kind of core concept in OpenMM.  There's a plugin that provides the integrators and DrudeForce.  As far as the core library is concern, they're just ordinary particles.  When a particle's mass is 0, that normally means it doesn't get integrated and the minimizer shouldn't move it.  So you need to give the Drude particles a small mass for minimizing, but then set them to be massless before integrating.

As you can tell, DrudeSCFIntegrator is a feature that's never gotten used very much.  It has some rough edges.  Drude simulations almost always use Langevin integration.  But I do want to make it work a bit better if I can.

**Comment by selimsami** [comment]
Aha.

When I set Drude mass to 0 and then do:
`simulation.context.setVelocitiesToTemperature(300*kelvin)`
all velocities, including atoms, are zero. 
But then:
`system.addForce(AndersenThermostat(300*kelvin, 1/picosecond))`
seems to work correctly - it gradually increases the temperature from 0 to 300.

So I guess you have to set DrudeMass to zero after the velocity initialization, that is very specific haha

**Comment by peastman** [comment]
That's not supposed to happen!  `setVelocitiesToTemperature()` should work normally.  Let me investigate that...

**Comment by peastman** [comment]
The fix is in #4356.

**Comment by selimsami** [comment]
No longer getting all zeros but I get a temperature of 600 K now when I set it to 300K.

**Comment by selimsami** [comment]
@peastman,
> I found another reason why it's so slow: the minimization tolerance is way too low. You set it to 0.01 kJ/mol/nm, which is far smaller than it can achieve. Even the default value 0.1 kJ/mol/nm is far too low. We should consider increasing it. The optimizer thrashes for a long time before finally giving up, usually with a RMS force on the Drude particles somewhere between 10 and 50 kJ/mol/nm. If you set the tolerance to 50 it gets much faster with no change in the actual accuracy.

I've been looking at this, I wonder how you got the forces of 10 and 50 kJ/mol/nm?

For the initial frame, I seem to be getting with my model:
The scalar RMS force on drude particles of  0.3 with CPU and 0.1 with Reference (even maximum force on a shell atom is 3 and 2 respectively)

On the other hand, with SWM4 force field, I am able to achieve:
Mean force on drude particles of ~6e-3 with CPU and ~2e-7 with Reference

I am a bit puzzled also by the fact that I dont see any improvement with Reference for my model. Any ideas why that could be?

I think for my model, the forces on Drude particles are too high to consider them 0 for dynamics... I was hoping the double precision would help though.

**Comment by selimsami** [comment]
Oh okay. Increasing EwaldErrorTolerance from 5e-4 to 2.5e-4 changed my model's:
The scalar RMS force on drude particles of ~0.2 with CPU and ~1e-7 with Reference 

I did not expect such a difference for the double precision case...
